### ISSUE TITLE
Fails to parse GADTs with strictness annotations

### ISSUE THREAD

**Issue by dcastro** [haskell/haddock#1053]
Haddock fails to parse when I place a comment on a data constructor field, whilst using GADT syntax and strictness annotations.

Using just GADT syntax, or just strictness annotations works, but I can't use both.

```haskell
data F a where
  X ::
      !Int -- ^ comment
   -> F Int
```
```
    • Unexpected strictness annotation: !Int
      strictness annotation cannot appear nested inside a type
    • In the type '!Int " comment"'
      In the definition of data constructor 'X'
      In the data declaration for 'F'
    |
105 |       !Int -- ^ comment
    |       ^^^^
```

For other people having the same issue as me, I'm working around it by using type equalities instead of GADT syntax:

```haskell
data F a
  = (a ~ Int) => X
    !Int -- ^ comment
```

Haddock version: 2.22.0
GHC: 8.6.3
Stack: 1.9.3
LTS: lts-13.7

**Comment by harpocrates** [comment]
Confusingly enough, this is actually a bug in the GHC parser, not Haddock.

Reported upstream: https://gitlab.haskell.org/ghc/ghc/issues/16585.

**Comment by harpocrates** [comment]
This is fixed in GHC `master` (See https://gitlab.haskell.org/ghc/ghc/merge_requests/794/diffs?commit_id=99dd5d6b8365ecc8748651395c503b2c0b82490e).

**Comment by harpocrates** [comment]
Closing as this is fixed in Haddock bundled with GHC 8.10.1.
